Функциональность «Рекомендации». 3 SP

SlavaTurovski opened the pull request.

SlavaTurovski committed:
Добавление функциональности «Рекомендации»

SlavaTurovski committed:
Правки импорта

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/repository/FilmRepository.java:
```diff
@@ -12,10 +12,7 @@
-import java.util.List;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
не особо принципиально. но лучше оставить весь набор (java.util...) а не звездочку 
<img width="1240" height="741" alt="image" src="https://github.com/user-attachments/assets/ced1c993-6590-487e-99dd-9d75ca5abbf4" />

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/repository/FilmRepository.java:
```diff
@@ -290,4 +287,39 @@ public List<Film> getFilmByDirector(Long directorId, String sortBy) {
     }
+
+    public List<Long> getUsersWithSameLikes(Long userId) {
+        final String sqlQuery = """
```
sql лучше вынести в константы. в начале класса (там где и остальные sql)

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/repository/FilmRepository.java:
```diff
@@ -290,4 +287,39 @@ public List<Film> getFilmByDirector(Long directorId, String sortBy) {
+    }
+
+    public List<Long> getFilmRecommendations(Long userId, List<Long> sameUserIds) {
+        String sqlQuery = """
```
тоже самое. sql необходимо вынести в константы

imaspa replied:
думаю в самом начале надо добавить ранний выход если пришел пустой similarUserIds
```
    if (similarUserIds.isEmpty()) {
        return Collections.emptyList(); 
    }
```

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/repository/FilmRepository.java:
```diff
@@ -290,4 +287,39 @@ public List<Film> getFilmByDirector(Long directorId, String sortBy) {
+                GROUP BY fl.user_id
+                having rate > 1
+                ORDER BY rate desc
+                limit 10
```
лимит лучше не зашивать, а пробрасывать через сигнатуру метода

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/repository/FilmRepository.java:
```diff
@@ -290,4 +287,39 @@ public List<Film> getFilmByDirector(Long directorId, String sortBy) {
+        final String sqlQuery = """
+                SELECT fl.user_id, count(fl.film_id) rate
+                FROM film_like ul
+                JOIN film_like fl ON (ul.film_id = fl.film_id AND ul.user_id != fl.user_id)
```
наверное лучше вынсти проверку в where
`AND fl.user_id != ul.user_id,`
а тут оставить
`JOIN film_like fl ON ul.film_id = fl.film_id`

imaspa commented on src/main/java/ru/yandex/practicum/filmorate/data/service/UserService.java:
```diff
@@ -110,4 +114,21 @@ public List<UserDto> getCommonFriends(Long userId, Long otherId) throws NotFound
+
+    public List<Film> getRecommendations(Long idUser) throws NotFoundException {
+        List<Long> sameUserIds = filmRepository.getUsersWithSameLikes(idUser);
+        log.debug("Получаем рекомендации для пользователя с ID {}", idUser);
```
для однотипности логов лучше придерживаться устоявшегося описания
вместо
`log.debug("Получаем рекомендации для пользователя с ID {}", idUser);`
указать
`log.debug("Рекомендации для пользователя  (старт). Id пользователя:  {}", idUser);`
по аналогии и для окончания

imaspa reviewed (commented):
Проверь пожалуйста на прохождение тестов (https://github.com/yandex-praktikum/java-filmorate/blob/develop/postman/sprint.json)
понятно что они упадут. т.к. это комплексные тесты, а у нас еще и половины нет, но там на add-recommendations написано 10 тестов, а в тестах ветки их всего 6

SlavaTurovski committed:
Правки после ревью

- исправил импорты
- в логах поменял описание
- перенес запросы в константы
- поправил запрос
- лимит 10 перенес
